### PR TITLE
Introductory Work - up to parallel processing of latitudes

### PR DESCRIPTION
**Pull request by lthUniBonn**
First attempt at a pull request - this should include my working solutions to #3 and #4. 

#3 is replaced mainly in the plot_maps (xarray requires slightly different array handling) 
- in the processing only the output formats are changed to netCDF4
- adaptive plotting is not included yet, so the axis scaling is still to be adapted manually, some of my first changes on the scaling are sadly also included here, but it still works just as before so I hope it's fine until I get to change the plotting fully  

#4 the parallel_processing.. uses command line arguments to chose the latitude subsets, output files are saved in /lats subfolder
- in principle also the previous handling of all latitudes one after the other is still included

From the beginning of my introduction a slight change in the readme is included, as one sentence gave some outdated info on the downloading process I think. 

Also the configuration changes for the now finishing 'Christmas Download' is included, which I also think should be good, as it is supposed to be the working configuration from now on. (e.g. grid size, number of downloaded levels, ..) 

I think I'll try to separate the work in more branches in the future. Any feedback is welcome!
Cheers, 
Lavinia

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -27,14 +27,16 @@
 # General settings.
 start_year = 2011
 final_year = 2017
-era5_data_dir = '~/ERA5Data/'
-model_level_file_name_format = 'ml_{:d}_{:02d}.netcdf'  # "{:d}_europe_{:d}_130_131_132_133_135.nc"
-surface_file_name_format = 'sfc_{:d}_{:02d}.netcdf'  #"{:d}_europe_{:d}_152.nc"
+era5_data_dir = '/cephfs/user/s6lathim/ERA5Data-112/'
```
Comment (markschelbergen) [inline config.py:30]: put back the old data dir and file names

File: config.py [hunk 2]
```diff
@@ -27,14 +27,16 @@
...
+model_level_file_name_format = "{:d}_europe_{:d}_130_131_132_133_135.nc"  # 'ml_{:d}_{:02d}.netcdf'
+surface_file_name_format = "{:d}_europe_{:d}_152.nc" # 'sfc_{:d}_{:02d}.netcdf'
```
Comment (markschelbergen) [inline config.py]: put back the old data dir and file names

File: config.py [hunk 3]
```diff
@@ -27,14 +27,16 @@
...
 # Downloading settings.
 area = "65/-20/30/20"
-upper_level = 122
+upper_level = 112 
 
 # Processing settings.
-output_file_name = "results/processed_data_{:d}_{:d}.nc".format(start_year, final_year)
-read_n_lats_at_once = 1
\ No newline at end of file
+output_file_name = "/cephfs/user/s6lathim/ERA5Data/results/lats/processed_data_europe_{:d}_{:d}.nc".format(start_year, final_year)
```
Comment (markschelbergen) [inline config.py]: put back the old data dir and file names

File: plot_maps.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 """
 from netCDF4 import Dataset
```
Comment (markschelbergen) [inline plot_maps.py]: netcdf4 library is not needed anymore, right?

File: parallel_process_data.py [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+# -*- coding: utf-8 -*-
+"""Process wind data for adjacent years and save processed data to a NetCDF file. Settings, such as target file name,
+are imported from config.py.
+
+Example::
+
+    $ python process_data.py
+
+"""
+from netCDF4 import Dataset
```
Comment (markschelbergen) [inline parallel_process_data.py]: Still uses netCDF4

File: process_data.py [hunk 6]
```diff
@@ -19,10 +19,10 @@
 from config import start_year, final_year, era5_data_dir, model_level_file_name_format, surface_file_name_format,\
     output_file_name, read_n_lats_at_once
 
-# Set the relevant heights for the different analysis types.
+# Set the relevant heights for the different analysis types in meter.#, 1500.],
```
Comment (markschelbergen) [inline process_data.py]: remove end of sentence

File: parallel_process_data.py [hunk 7]
```diff
@@ -0,0 +1,600 @@
+#!/usr/bin/env python2
```
Comment (markschelbergen) [inline parallel_process_data.py]: Why have a new file, seems like it copies almost all the functionalities in the process_data.py? If so, import it the function from there. We shouldn't maintain pieces of code that are the same in two separate files.

File: process_data.py [hunk 8]
```diff
@@ -370,7 +369,7 @@ def write_results_to_output_netcdf(output_variables, result_arrays):
 
 def create_and_configure_output_netcdf(hours, lats, lons, output_file):
     # Write output to a new NetCDF file.
-    nc_out = Dataset(output_file, "w", format="NETCDF3_64BIT_OFFSET")
+    nc_out = Dataset(output_file, "w", format="NETCDF4")
```
Comment (markschelbergen) [inline process_data.py:430]: I recall that I had some issues before with the default format. Is suppose that would be the NETCDF4 format that you are suggesting now. Do you have an idea why NETCDF3 was used here previously and did you test the new format?

File: plot_maps.py [hunk 9]
```diff
@@ -415,7 +450,7 @@ def percentile_plots_ref(plot_var, i_case, plot_var_ref, i_case_ref, plot_settin
 
 def plot_figure5():
     """" Generate integrated mean power plot. """
-    column_titles = ["50 - 150m", "0 - 10km", "Ratio"]
+    column_titles = ["50 - 150m", "10 - 500m", "Ratio"]#"0 - 10km", "Ratio"]
```
Comment (markschelbergen) [inline plot_maps.py]: Remove commented out code at end of line

File: plot_maps.py [hunk 10]
```diff
@@ -427,23 +462,33 @@ def plot_figure5():
...
+        #'colorbar_tick_fmt': '{:.0f}',
+        'contour_line_levels': [10, 17],
+        'contour_fill_levels': np.logspace(np.log10(5), np.log10(20.0), num=16),
+        'colorbar_ticks': [5, 10, 15, 20],
         'colorbar_tick_fmt': '{:.0f}',
         'colorbar_label': 'Increase factor [-]',
     }
 
     plot_items = [plot_item0, plot_item1, plot_item2]
+    print(plot_item1['data']/plot_item0['data'])
```
Comment (markschelbergen) [inline plot_maps.py]: Remove

File: plot_maps.py [hunk 11]
```diff
@@ -427,23 +462,33 @@ def plot_figure5():
...
     plot_item2 = {
         'data': plot_item1['data']/plot_item0['data'],
         'log_scale': True,
-        'contour_line_levels': [300., 600.],
-        'contour_fill_levels': np.logspace(np.log10(100.0), np.log10(6700.0), num=16),
-        'colorbar_ticks': [100., 300., 1000., 3000.],
+        #'contour_line_levels': [300., 600.],
+        #'contour_fill_levels': np.logspace(np.log10(100.0), np.log10(6700.0), num=16),
+        #'colorbar_ticks': [100., 300., 1000., 3000.],
+        #'colorbar_tick_fmt': '{:.0f}',
```
Comment (markschelbergen) [inline plot_maps.py]: Remove

File: plot_maps.py [hunk 12]
```diff
@@ -427,23 +462,33 @@ def plot_figure5():
...
     plot_item1 = {
         'data': p_integral_mean[1, :, :]*1e-6,
-        'contour_line_levels': np.linspace(0, 65, 21)[::4],
-        'contour_fill_levels': np.linspace(0, 65, 21),
-        'colorbar_ticks': np.linspace(0, 65, 21)[::4],
-        'colorbar_tick_fmt': '{:.0f}',
+        #'contour_line_levels': np.linspace(0, 65, 21)[::4],
+        #'contour_fill_levels': np.linspace(0, 65, 21),
+        #'colorbar_ticks': np.linspace(0, 65, 21)[::4],
+        #'colorbar_tick_fmt': '{:.0f}',
```
Comment (markschelbergen) [inline plot_maps.py]: Remove

File: plot_maps.py [hunk 13]
```diff
@@ -31,17 +34,49 @@
...
+            print('All latitudes are read from files with the respective latitude in a similar fashion to: ' + resources[0])
+            nc = xr.open_mfdataset(resources, concat_dim='latitude')
+else:
+    nc = xr.open_dataset(output_file_name)
+
+#* change to xarray!
+lons = nc['longitude'].values
+lats = nc['latitude'].values
+print(lons)
+print(lats)
```
Comment (markschelbergen) [inline plot_maps.py]: remove

File: plot_maps.py [hunk 14]
```diff
@@ -31,17 +34,49 @@
...
+
+            # sorting of provided latitudes included
+            lats = [f.split('_')[-1][:-(len(f.split('.')[-1])+1)] for f in os.listdir(output_dir) if f.split("/")[-1].split("lats")[0] == output_file_name_prefix]
+            lats.sort(key=float) 
+            print(str(len(lats)) + ' latitudes found in directory ' + output_dir + ' for the years ' + str(start_year) + ' to ' + str(final_year) + ':')
+            print(lats)
+
+            resources = [output_dir + output_file_name_prefix + 'lats_' + lat + '.nc' for lat in lats]
+            print('All latitudes are read from files with the respective latitude in a similar fashion to: ' + resources[0])
+            nc = xr.open_mfdataset(resources, concat_dim='latitude')
```
Comment (markschelbergen) [inline plot_maps.py]: Please add more comments, because it's hard to follow what exactly you are doing here.

File: parallel_process_data.py [hunk 15]
```diff
@@ -0,0 +1,600 @@
...
+        except getopt.GetoptError:
+            print ("parallel_process_data.py -s SubsetNumber >> process individual latitude subset SubsetNumber \n -h >> display help ")
+            sys.exit()
+        for opt, arg in opts:
+            if opt in ("-h", "--help"):
+                print ("parallel_process_data.py -s SubsetNumber >> process individual latitude subset SubsetNumber \n -h >> display help ")
+                sys.exit()
+            elif opt in ("-s", "--subset"):
+                # Select only a specific latitude subset
+                subset_number_input = int(arg)
```
Comment (markschelbergen) [inline parallel_process_data.py]: Please add more comments explaining how the code works.

File: parallel_process_data.py [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+
+    print('Locations analyzed: ({}/{}).'.format(counter, total_iters))
+    time_lapsed = float(timer()-start_time)
+    time_remaining = time_lapsed/counter*(total_iters-counter)
+    print("Time lapsed: {:.2f} hrs, expected time remaining: {:.2f} hrs.".format(time_lapsed/3600,
+                                                                                     time_remaining/3600))
+    return(res, counter)
+
+
+def process_complete_grid(output_file, subset_number_input):
```
Comment (markschelbergen) [inline parallel_process_data.py]: the function is called complete grid, but you have a new argument 'subset_number_input'. This suggests that the function is not for the full grid.

File: parallel_process_data.py [hunk 17]
```diff
@@ -0,0 +1,600 @@
...
+    n_subsets = int(np.ceil(float(len(lats)) / read_n_lats_at_once))
+
+    if subset_number_input == (-1):
+        i_subset = 0
+    elif subset_number_input >= n_subsets:
+        raise ValueError("User input subset number ({:.0f}) larger than total maximal subset index ({:.0f}) starting at 0.".format(subset_number_input, (n_subsets-1)))
+    else:
+        # User input given - processing only one latitude subset:
+        i_subset = subset_number_input
+        total_iters = len(lons)
```
Comment (markschelbergen) [inline parallel_process_data.py]: take this outside this function

File: parallel_process_data.py [hunk 18]
```diff
@@ -0,0 +1,600 @@
...
+    gaps = d_hours != 1
+    if np.any(gaps):
+        i_gaps = np.argwhere(gaps)
+        for i in i_gaps:
+            print("Gap found between {} and {}.".format(hour_to_date_str(hours[i]),
+                                                        hour_to_date_str(hours[i+1])))
+
+
+
+def process_lats_subset(lats_subset, lenLons, levels, lenHours, v_levels_east, v_levels_north, v_levels, t_levels, q_levels, surface_pressure, heights_of_interest, analyzed_heights_ids, res, start_time, counter, total_iters):
```
Comment (markschelbergen) [inline parallel_process_data.py]: please rewrite such that each parallel process outputs a separate netcdf, which we concatenate in a post processing step

**Review by markschelbergen — CHANGES_REQUESTED** [review]
I don't understand the flow for parallel processing. I provided comments on the pieces of codes that I have doubts about. However, let's discuss in a Skype call such that you can explain me your rational better.

File: config.py (hunk 2 above)
Comment (lthUniBonn) [reply config.py]: taken care of - commit e3b9ba8

File: config.py (hunk 3 above)
Comment (lthUniBonn) [reply config.py]: taken care of (same as above) - commit e3b9ba8

File: plot_maps.py (hunk 4 above)
Comment (lthUniBonn) [reply plot_maps.py]: Yes, forgot to remove that, it is included in xarray. 
Removed in commit 12e3200

File: process_data.py (hunk 6 above)
Comment (lthUniBonn) [reply process_data.py]: removed the second comment there, see commit 76e0cf4

File: parallel_process_data.py (hunk 18 above)
Comment (lthUniBonn) [reply parallel_process_data.py]: This is already happening, if read_n_lats_at_once is 1 - I had though to keep that variable, but at the moment if its larger, files span multiple latitudes - Now I removed the dependency on read_n_lats_at_once and now just process all latitudes individually 
54637c5 & 5dd45a1
Not sure if that's what you wanted though, or if its that best solution - so that we could talk about in Sykpe then.

File: parallel_process_data.py (hunk 16 above)
Comment (lthUniBonn) [reply parallel_process_data.py]: For better description the subset interpratation was replaced in favor of a latitude & latitude index naming 
in commit 68d38e1
Also the function naming has been replaced by more accurate job descriptions

File: parallel_process_data.py (hunk 17 above)
Comment (lthUniBonn) [reply parallel_process_data.py]: not sure why you ask that, I guess it might also be resolved by better function naming introduced in 68d38e1.

File: parallel_process_data.py (hunk 15 above)
Comment (lthUniBonn) [reply parallel_process_data.py]: This part is conderned with reading the user input, I included a bit more info on what exactly the reading does here, hope it is more clear now (in addition in the first lines I included more detailed instructions as well) via commit fbb2624
Your questions seem to show also that the name of the program is quite misleading, which I changed as well (4963f4e and in the file 084657a)

File: parallel_process_data.py (hunk 7 above)
Comment (lthUniBonn) [reply parallel_process_data.py]: i agree, but don't know which way to take best, for the small functions which are just a 1 on 1 copy I can easily do that, but not the processing / output files itself - so I'll wait with this for now until our meeting

File: parallel_process_data.py (hunk 5 above)
Comment (lthUniBonn) [reply parallel_process_data.py]: yes it does, to write the output - do you think it would be better to change all of that to xarray, too? 

That would mean to change the handling of the res dictionary and I think also the whole definition of the output_variables dimensions and so on to using an xarray dataset 
then this dataset would still be written to a netcdf4 file 

so I think the only merit would be a possibly prettier coding, as the res dictionary could maybe be modified to be an xarray dataset, then just write that...? I'm still getting into the use of xarray in this case, but I think that would be the workflow

File: plot_maps.py (hunk 14 above)
Comment (lthUniBonn) [reply plot_maps.py]: changed to subsets and greatly simplified by using one set input parameter, commit 141ec71

File: plot_maps.py (hunk 9 above)
Comment (lthUniBonn) [reply plot_maps.py]: all unnecessary comments/print statements removed with commit 558e5fb
